r/src/array.cpp

```diff
@@ -601,22 +623,51 @@ SEXP Array__as_vector(const std::shared_ptr<arrow::Array>& array) {
   using namespace arrow::r;
 
   switch (array->type_id()) {
-    case Type::BOOL:
-      return BooleanArray_to_Vector(array);
+    // direct support
     case Type::INT8:
       return simple_Array_to_Vector<RAWSXP>(array);
     case Type::INT32:
       return simple_Array_to_Vector<INTSXP>(array);
     case Type::DOUBLE:
       return simple_Array_to_Vector<REALSXP>(array);
+
+    // need to handle 1-bit case
+    case Type::BOOL:
+      return BooleanArray_to_Vector(array);
+
+    // handle memory dense strings
     case Type::STRING:
       return StringArray_to_Vector(array);
     case Type::DICTIONARY:
       return DictionaryArray_to_Vector(static_cast<arrow::DictionaryArray*>(array.get()));
+
     case Type::DATE32:
       return Date32Array_to_Vector(array);
     case Type::DATE64:
       return Date64Array_to_Vector(array);
+
+    // promotions to integer vector
+    case Type::UINT8:
+      return arrow::r::promotion_Array_to_Vector<INTSXP, arrow::UInt8Type>(array);
+    case Type::INT16:
+      return arrow::r::promotion_Array_to_Vector<INTSXP, arrow::Int16Type>(array);
+    case Type::UINT16:
+      return arrow::r::promotion_Array_to_Vector<INTSXP, arrow::UInt16Type>(array);
+
+    // promotions to numeric vector
+    case Type::UINT32:
+      return arrow::r::promotion_Array_to_Vector<REALSXP, arrow::UInt32Type>(array);
+    case Type::HALF_FLOAT:
+      return arrow::r::promotion_Array_to_Vector<REALSXP, arrow::UInt32Type>(array);
+    case Type::FLOAT:
+      return arrow::r::promotion_Array_to_Vector<REALSXP, arrow::UInt32Type>(array);
+
+    // lossy promotions to numeric vector
+    case Type::INT64:
+      return arrow::r::promotion_Array_to_Vector<REALSXP, arrow::Int64Type>(array);
+    case Type::UINT64:
+      return arrow::r::promotion_Array_to_Vector<REALSXP, arrow::UInt64Type>(array);
```

Comment:
Should these error on data loss? opened https://issues.apache.org/jira/browse/ARROW-3553

Reply:
It should. However, I'll probably switch to using the bit64 package in a follow up pr. it just uses a numeric vector (array of double as the host and interpret the bytes as a `int64_t`. 

This way, there is no loss.